File: solutions/0861. Score After Flipping Matrix/0861_1.py
# Time Complexity: O(mn)
# Space Complexity: O(m)
class Solution:
  def matrixScore(self, grid: List[List[int]]) -> int:
    for row in grid:
      if row[0] == 0:
        self._flip(row)

    for j, col in enumerate(list(zip(*grid))):
      if sum(col) * 2 < len(grid):
        self._flipCol(grid, j)

    return sum(self._binary(row) for row in grid)

  def _flip(self, row: List[int]) -> None:
    for i in range(len(row)):
      row[i] ^= 1

  def _flipCol(self, grid: List[List[int]], j: int) -> None:
    for i in range(len(grid)):
      grid[i][j] ^= 1

  def _binary(self, row: List[int]) -> int:
    res = row[0]
    for j in range(1, len(row)):
      res = res * 2 + row[j]

    return res

File: solutions/0861. Score After Flipping Matrix/0861_1.rs
// Time Complexity: O(mn)
// Space Complexity: O(1)
impl Solution {
    pub fn matrix_score(mut grid: Vec<Vec<i32>>) -> i32 {
        let rows = grid.len();
        let cols = grid[0].len();

        for row in grid.iter_mut() {
            if row[0] == 0 {
                Self::flip(row);
            }
        }

        for j in 0..cols {
            let mut sum = 0;
            for i in 0..rows {
                sum += grid[i][j];
            }
            if sum * 2 < rows as i32 {
                Self::flip_col(&mut grid, j);
            }
        }

        grid.into_iter().map(Self::binary).sum()
    }

    fn flip(row: &mut Vec<i32>) {
        for val in row.iter_mut() {
            *val ^= 1;
        }
    }

    fn flip_col(grid: &mut Vec<Vec<i32>>, j: usize) {
        for row in grid.iter_mut() {
            row[j] ^= 1;
        }
    }

    fn binary(row: Vec<i32>) -> i32 {
        row.into_iter().fold(0, |acc, b| acc * 2 + b)
    }
}
